#2853: Member domains submit - [RJM]

rachidatecs opened the pull request:
## Ticket

<!-- PR title format: `#issue_number: Descriptive name ideally matching ticket name - [sandbox]`-->
Resolves #2853 

## Changes

<!-- What was added, updated, or removed in this PR. -->
- The review button toggles between the ajax table and a rendered read only view with a save button
- The breadcrumbs reflect the router config
- Save triggers a form submit with the removed and added domains
- Backend ajax endpoint
- Instead of deleting DomainInvitations, we move their status to CANCELED (new AC)
- Instead of adding DomainIncitations, we first do a lookup and if the invite exists we change its status to INVITED, otherwise we do a create (new AC)
- On success response, trigger a render of the readonly domains list with a trigger that activates a generic success message: The domain assignment changes have been saved. (redirect in JS after we get the ajax response)
- On error response, trigger a render of the edit page with a generic error message
- Clicking "go back" takes you to the "Edit domain assignments page"
- Unit tests

<!--
    Please add/remove/edit any of the template below to fit the needs
    of this specific PR.
--->

## Context for reviewers

<!--Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.  -->

## Setup

<!--  Add any steps or code to run in this section to help others run your code.
    
    Example 1:
    ```sh
    echo "Code goes here"
    ``` 
    
    Example 2: If the PR was to add a new link with a redirect, this section could simply be:
    -go to /path/to/start/page
    -click the blue link in the <insert location>
    -notice user is redirected to <proper end location>
-->

Make sure you test on a portfolio member and on an invited 'user'.
For the invited user, test that an existing invite flips between CANCELED and INVITED (a non-existing will get created).

## Code Review Verification Steps

### As the original developer, I have

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)
<!-- Mark "- N/A" and check at the end of each check that is not applicable to your PR -->
- [x] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [x] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [x] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

### As a code reviewer, I have

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Verified code meets all checks above. Address any checks that are not satisfied
- [x] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [x] Checked that all code is adequately covered by tests
- [x] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [x] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Meets all designs and user flows provided by design/product
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] (Rarely needed) Tested as both an analyst and applicant user

### As a designer reviewer, I have

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior. Comment any found issues or broken flows.
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

## Screenshots

<!-- If this PR makes visible interface changes, an image of the finished interface can help reviewers
and casual observers understand the context of the changes.
A before image is optional and can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow.
You may want to use [GIPHY Capture](https://giphy.com/apps/giphycapture) for this! 📸

_Please frame images to show useful context but also highlight the affected regions._
--->

abroddrick commented:
After editing a users domain I got this when going back to the members table:
![image](https://github.com/user-attachments/assets/a80003ca-1f22-40b1-8102-07a04cc8960f)
steps I took:

1. I created a new invitation (admin with the highest permissions)
2. I navigated to the Member Profile page for that invitation
3. I clicked the manage icon to edited their domains 
4. Added 8 domains to the user successfully. 
5. From the Member Profile page, I then click the "Members" breadcrumb to go back to the members table
6. saw the table was now empty when it was full before


The very ugly error logs: 
15:38:47.002: [APP/PROC/WEB.0] {"server_time": "26/Dec/2024 23:38:47,001", "level": "ERROR", "message": "Internal Server Error: /get-portfolio-members-json/\nTraceback (most recent call last):\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py\", line 89, in _execute\n    return self.cursor.execute(sql, params)\npsycopg2.errors.CardinalityViolation: more than one row returned by a subquery used as an expression\n\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/handlers/exception.py\", line 55, in inner\n    response = get_response(request)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/handlers/base.py\", line 197, in _get_response\n    response = wrapped_callback(request, *callback_args, **callback_kwargs)\n  File \"/home/vcap/deps/0/python/lib/python3.10/contextlib.py\", line 79, in inner\n    return func(*args, **kwds)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/views/generic/base.py\", line 104, in view\n    return self.dispatch(request, *args, **kwargs)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/contrib/auth/mixins.py\", line 109, in dispatch\n    return super().dispatch(request, *args, **kwargs)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/views/generic/base.py\", line 143, in dispatch\n    return handler(request, *args, **kwargs)\n  File \"/home/vcap/app/registrar/views/portfolio_members_json.py\", line 40, in get\n    page_obj = paginator.get_page(page_number)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py\", line 63, in get_page\n    number = self.validate_number(number)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py\", line 53, in validate_number\n    if number > self.num_pages:\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/utils/functional.py\", line 57, in __get__\n    res = instance.__dict__[self.name] = self.func(instance)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py\", line 99, in num_pages\n    if self.count == 0 and not self.allow_empty_first_page:\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/utils/functional.py\", line 57, in __get__\n    res = instance.__dict__[self.name] = self.func(instance)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py\", line 93, in count\n    return c()\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/query.py\", line 608, in count\n    return self.query.get_count(using=self.db)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/query.py\", line 568, in get_count\n    return obj.get_aggregation(using, {\"__count\": Count(\"*\")})[\"__count\"]\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/query.py\", line 554, in get_aggregation\n    result = compiler.execute_sql(SINGLE)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/compiler.py\", line 1562, in execute_sql\n    cursor.execute(sql, params)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py\", line 67, in execute\n    return self._execute_with_wrappers(\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py\", line 80, in _execute_with_wrappers\n    return executor(sql, params, many, context)\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py\", line 84, in _execute\n    with self.db.wrap_database_errors:\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/utils.py\", line 91, in __exit__\n    raise dj_exc_value.with_traceback(traceback) from exc_value\n  File \"/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py\", line 89, in _execute\n    return self.cursor.execute(sql, params)\ndjango.db.utils.ProgrammingError: more than one row returned by a subquery used as an expression\n"}

abroddrick commented:
**change request** When adding domains to a user who has none, it looks like the page still has a search bar and the text may not line up with [the figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8901-83619&t=rogV9cjqiYMjF4Zz-0) for this state. Could you change it to line up with that figma, if not no worries we can add that to another ticket.

abroddrick commented:
**BUG #2**
I switched the portfolio so I could see the members table again and got a weird bug where I couldn't cancel the changes I made to a domain and could only save. The steps were as follows:

1. created a new invitation like before
2. navigated to that invitation's member profile page
3. clicked manage to edit domains
4. selected a bunch of domains
5. clicked cancel- nothing happened
6. tried to navigate with tab and select via keyboard and still it didn't work
(note clicking save worked though)


console log: 
![image](https://github.com/user-attachments/assets/cd61cc4e-8ca5-4da2-b504-dd298c9f1852)

abroddrick commented:
@rachidatecs I saw your question on the ticket and to avoid convos in two places, I figured I would reply here. We have a follow on ticket ([here-2960](https://github.com/cisagov/manage.get.gov/issues/2960)) to change all email sending for domain invitations so that if there are multiple domain invitations sent at once it instead sends a combined email domain invitation email that just lists all the emails. The assumption was this ticket would send emails inviting people to the domain. I think when the user clicks the final save it should send domain invitation emails. (However, I wouldn't be opposed to us just expanding that 2960 ticket to just include adding emails to this workflow)

abroddrick commented:
**possible bug 3 (or issue with my local)**

I went to edit the assigned domains for a domain manager who is already a member and has domains and my local gave me a 500 error saying "django.template.exceptions.TemplateSyntaxError: Unclosed tag on line 8: 'block'. Looking for one of: endblock."

abroddrick commented on src/registrar/assets/src/js/getgov/table-edit-member-domains.js:
```diff
@@ -217,7 +228,125 @@ export class EditMemberDomainsTable extends BaseTable {
+      console.warn('Missing DOM element. Expected element with id review-domain-assignments')
+    }
+
+    if (this.backButton.length == 2) {
```
**non blocking question (also happy to just huddle if chatting is faster)** this is the error I was seeing, can you explain what's happening here a bit more?

rachidatecs replied:
Yeah, we started adding JS console warnings when we make a string assumption that an element with a certain ID/class exists and we add some event listeners to it. If that element does not exist, the JS will error out and the stack will stop running. The console warnings basically do the job of a try/catch

We only started trying this out, not sure whether this is a practice that's good enough to adopt everywhere we're adding event listeners or doing actions on DOM elements.

abroddrick replied:
That makes sense, resolving.

abroddrick commented on src/registrar/assets/src/sass/_theme/_typography.scss:
```diff
@@ -23,6 +23,14 @@ h2 {
   color: color('primary-darker');
 }
 
+.header--body {
```
**non blocking question** can you explain the switch to header-- body?

rachidatecs replied:
Ah yes. The old CSS class identified those style headers as a register form header, semantically (screen readers will read classes I think): register-form-review-header

The Figma for this ticket introduces a header variant on pages other than the register form that look similar to those old headers on the register form. SO, I created a more generic CSS class called header--body that indicates that the header is a variation of type --body (the font size matches the `<p>` font size. Not the best CSS class name but I couldn't think of something better.

abroddrick replied:
Oh that makes sense! Thanks for the explanation, I wouldn't have guessed that otherwise! resolving this convo.

abroddrick commented on src/registrar/assets/src/sass/_theme/_typography.scss:
```diff
@@ -23,6 +23,14 @@ h2 {
+  margin-top: units(2);
+  font-weight: font-weight('semibold');
+  // The units mixin can only get us close, so it's between 
+  // hardcoding the value and using  in markup
```
**the most nitpickiest of nitpicks**
```suggestion
  // hardcoding the value and using in markup
```

abroddrick commented on src/registrar/templates/includes/request_review_steps.html:
```diff
@@ -132,7 +132,7 @@ <h3 class="register-form-review-header">Alternative domains</h3>
             {% if domain_request.has_additional_details %}
                 {% include "includes/summary_item.html" with title="Additional Details" value=" " heading_level=heading_level editable=is_editable edit_link=domain_request_url %}
-                <h3 class="register-form-review-header">CISA Regional Representative</h3>
+                <h3 class="header--body text-primary-dark margin-bottom-0">CISA Regional Representative</h3>
```
**suggestion:** there are 10 places using this same css in this PR alone, how about instead we either update header--body or just make a new css definition that extends header--body and include the text dark and 0 bottom margin customizations. That way if we want to change this down the road we aren't changing it in 10 places

rachidatecs replied:
See my comment above. Design introduced a new header variant, so we now have 2 types of headers that share the same CSS (size, top margin) but differ on bottom margin and color. I believe the common CSS should be defined in the common class (header--body) and the variations on bottom margin and color should either be defined in a secondary class or be applied through the utility classes as done here.

Ideally - and this should be posed to @cisagov/gov-designers - we do not need such minor variations on color and margins and we can get rid of the utility classes + rename header--body or just apply the styles of header--body directly to `h4` and a class called `.h4 `

Katherine-Osos replied:
Agree that small variations should be avoided. cc: @gabydisarli for input

abroddrick replied:
I think @Katherine-Osos and @gabydisarli this really applies to the new ticket for auditing all of our styles/ headings and creating a more consistent solution. Definitely counting this out of scope for this PR.

abroddrick commented on src/registrar/views/portfolios.py:
```diff
@@ -341,6 +407,91 @@ def get(self, request, pk):
+                messages.error(request, "A database error occurred while saving changes.")
+                return redirect(reverse("invitedmember-domains-edit", kwargs={"pk": pk}))
+            except Exception as e:
+                messages.error(request, f"An unexpected error occurred: {str(e)}")
```
**change before merging:** please log any error that occurs in the exceptions( this applies to above here as well)

Also let's change this to somewhat match what we have for epp errors where we say to message help@get.gov if the error persists. I'm also a tad unsure about showing the user the error type we get, but I could see that being helpful to help find their issue in the logs.

rachidatecs replied:
I can log the error type and change the error message as you describe

abroddrick commented on src/registrar/views/portfolios.py:
```diff
@@ -341,6 +407,91 @@ def get(self, request, pk):
+                messages.error(request, f"An unexpected error occurred: {str(e)}")
+                return redirect(reverse("invitedmember-domains-edit", kwargs={"pk": pk}))
+        else:
+            messages.info(request, "No changes detected.")
```
**possibly blocking** I would prefer that we just show this if a user clicks the review button without having clicked anything. It seems very odd they can get get to the review page with no changes. Would this be too much of an addition to add in?

rachidatecs replied:
I think this is another question for design. We typically do not limit the interaction with submit buttons, but rather let the user hit submit and respond to whatever selection/ errors / no selections they have made. Our implementation follows that pattern. FWIW, I think it's the right pattern.

Katherine-Osos replied:
Agree that it follows our current pattern, which doesn't limit submission if no changes were made. I'd suggest keeping it consistent for now, but maybe this is something we could circle back to later - across registrar forms. cc: @gabydisarli for input

abroddrick replied:
Sounds good, resolving this convo

abroddrick commented on src/registrar/views/portfolios.py:
```diff
@@ -341,6 +407,91 @@ def get(self, request, pk):
+        try:
+            return json.loads(domain_data) if domain_data else []
+        except json.JSONDecodeError:
+            messages.error(self.request, f"Invalid data for {domain_type}.")
```
**change before merging** log the error and also let's have the message tell the user to email if this error persists (like I mention above)

abroddrick commented on src/registrar/views/portfolios.py:
```diff
@@ -216,6 +220,68 @@ def get(self, request, pk):
+                messages.error(request, "A database error occurred while saving changes.")
+                return redirect(reverse("member-domains-edit", kwargs={"pk": pk}))
+            except Exception as e:
+                messages.error(request, f"An unexpected error occurred: {str(e)}")
```
**change before merging** same comments in this block that I mention below, log errors and add language so that a user emails help@get.gov if the error persists.

abroddrick reviewed (commented):
Submitting my review so far, some changes requested and other questions. Will review more once it's not buggy on local and sandbox.

rachidatecs commented:
> **BUG**
> 
> After editing a users domain I got this when going back to the members table: ![image](https://private-user-images.githubusercontent.com/109625347/398815131-a80003ca-1f22-40b1-8102-07a04cc8960f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1OTM5NTQsIm5iZiI6MTczNTU5MzY1NCwicGF0aCI6Ii8xMDk2MjUzNDcvMzk4ODE1MTMxLWE4MDAwM2NhLTFmMjItNDBiMS04MTAyLTA3YTA0Y2M4OTYwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIzMFQyMTIwNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjczYzI2OWMzZWE0NzY1ZGRiYzA4NzYxZWQ4ZDRiNzZmNzAyYzE2ZWUxOWQwMDg1MTIzNjE2NzE3YzAzYTUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5SkB6j5fC1rX_r7tkOF61ABA1vmkr_rbRYDWqO0sk50) steps I took:
> 
> ...

On that first bug - this was an issue with `initial_invitations_search` in portfolio_members_json where a subquery (`Subquery(domain_invitations.values("domain_info")`) would break if thee were more than one row returned. I still do not fully understand it but I have a fix for you and Dave to review (this goes deep into how the abstracted SQL query works).

dave-kennedy-ecs commented:
> **BUG**
> 
> After editing a users domain I got this when going back to the members table: ![image](https://private-user-images.githubusercontent.com/109625347/398815131-a80003ca-1f22-40b1-8102-07a04cc8960f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU2MDY5NTgsIm5iZiI6MTczNTYwNjY1OCwicGF0aCI6Ii8xMDk2MjUzNDcvMzk4ODE1MTMxLWE4MDAwM2NhLTFmMjItNDBiMS04MTAyLTA3YTA0Y2M4OTYwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIzMVQwMDU3MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNWExZjVhZmNhYmY2OWNhMmM1Mjk2Y2QyZTBlY2FkMThkYzFjMTI1OWJjOGU1OTE0MjE5M2E1ZGFlMGU2MDA0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CTG8CVvBhGBJPaVjncAj69htihLrOru2mOPvWvX1TYQ) steps I took:
> 
> 1. I created a new invitation (admin with the highest permissions)
> 2. I navigated to the Member Profile page for that invitation
> 3. I clicked the manage icon to edited their domains
> 4. Added 8 domains to the user successfully.
> 5. From the Member Profile page, I then click the "Members" breadcrumb to go back to the members table
> 6. saw the table was now empty when it was full before
> 
> The very ugly error logs: 15:38:47.002: [APP/PROC/WEB.0] {"server_time": "26/Dec/2024 23:38:47,001", "level": "ERROR", "message": "Internal Server Error: /get-portfolio-members-json/\nTraceback (most recent call last):\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute\n return self.cursor.execute(sql, params)\npsycopg2.errors.CardinalityViolation: more than one row returned by a subquery used as an expression\n\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner\n response = get_response(request)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response\n response = wrapped_callback(request, *callback_args, **callback_kwargs)\n File "/home/vcap/deps/0/python/lib/python3.10/contextlib.py", line 79, in inner\n return func(*args, **kwds)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/views/generic/base.py", line 104, in view\n return self.dispatch(request, *args, **kwargs)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/contrib/auth/mixins.py", line 109, in dispatch\n return super().dispatch(request, *args, **kwargs)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/views/generic/base.py", line 143, in dispatch\n return handler(request, _args, **kwargs)\n File "/home/vcap/app/registrar/views/portfolio_members_json.py", line 40, in get\n page_obj = paginator.get_page(page_number)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py", line 63, in get_page\n number = self.validate_number(number)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py", line 53, in validate_number\n if number > self.num_pages:\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/utils/functional.py", line 57, in **get**\n res = instance.**dict**[self.name] = self.func(instance)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py", line 99, in num_pages\n if self.count == 0 and not self.allow_empty_first_page:\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/utils/functional.py", line 57, in **get**\n res = instance.**dict**[self.name] = self.func(instance)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/core/paginator.py", line 93, in count\n return c()\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/query.py", line 608, in count\n return self.query.get_count(using=self.db)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/query.py", line 568, in get_count\n return obj.get_aggregation(using, {"__count": Count("_")})["__count"]\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/query.py", line 554, in get_aggregation\n result = compiler.execute_sql(SINGLE)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1562, in execute_sql\n cursor.execute(sql, params)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 67, in execute\n return self._execute_with_wrappers(\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers\n return executor(sql, params, many, context)\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute\n with self.db.wrap_database_errors:\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/utils.py", line 91, in **exit**\n raise dj_exc_value.with_traceback(traceback) from exc_value\n File "/home/vcap/deps/0/python/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute\n return self.cursor.execute(sql, params)\ndjango.db.utils.ProgrammingError: more than one row returned by a subquery used as an expression\n"}

Confirmed this bug. And also that it was already present before our ticket. Can we tackle this one as a separate bug?

rachidatecs commented:
See if this commit helps: 1547a107eaad45509b822e536c96a730573a7eb7

We can still follow through with cleanup if you think this is too messy. Maybe a unit test too.

rachidatecs commented:
> **change request** When adding domains to a user who has none, it looks like the page still has a search bar and the text may not line up with [the figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8901-83619&t=rogV9cjqiYMjF4Zz-0) for this state. Could you change it to line up with that figma, if not no worries we can add that to another ticket.

Solved

rachidatecs commented:
> **BUG #2** I switched the portfolio so I could see the members table again and got a weird bug where I couldn't cancel the changes I made to a domain and could only save. The steps were as follows:
> 
> ...

Fixed. I had different breadcrumbs that included a link to trigger the modal, but then revised. Thanks for catching this.

rachidatecs commented:
> @rachidatecs I saw your question on the ticket and to avoid convos in two places, I figured I would reply here. We have a follow on ticket ([here-2960](https://github.com/cisagov/manage.get.gov/issues/2960)) to change all email sending for domain invitations so that if there are multiple domain invitations sent at once it instead sends a combined email domain invitation email that just lists all the emails. The assumption was this ticket would send emails inviting people to the domain. I think when the user clicks the final save it should send domain invitation emails. (However, I wouldn't be opposed to us just expanding that 2960 ticket to just include adding emails to this workflow)

Yeah, emails are definitely a follow-on to this ticket (2853) so they can be implemented cleanly after we revise to implement email concatenation.

rachidatecs commented:
> **possible bug 3 (or issue with my local)**
> 
> I went to edit the assigned domains for a domain manager who is already a member and has domains and my local gave me a 500 error saying "django.template.exceptions.TemplateSyntaxError: Unclosed tag on line 8: 'block'. Looking for one of: endblock."
> 
> **possible bug 4 (or issue with my local)** I tried running this locally and adding a domains to an invitation that had 0 domains to start with, however, after clicking on two domains, I couldn't navigate to the review page. It was just stuck. I clicked cancel and then tried again and the second time it allowed me get to the review page with no issue.

Unable to recreate possible bugs #3 and #4

abroddrick commented:
**minor bug:** 
When you add domains to an existing member it doesn't fill in the "role" value on UserDomainRoles (though granted we may just want to chop off that field later, given we have yet to add another role into the mix). 
Please have the role value filled in before merging, just for consistency. 
![image](https://github.com/user-attachments/assets/d7fd9d50-fc19-45ff-a66a-bbebdef334b2)

abroddrick reviewed (approved):
**minor bug**
 I don't _think_ this is from your code, but there is a bug.

Looks like I am assigned 8 domains but these two areas show 1 domain assigned: 
![Screenshot 2024-12-31 at 3 09 25 PM](https://github.com/user-attachments/assets/89464444-ad75-4b96-8f45-b8a6afdf7c0c)
![Screenshot 2024-12-31 at 3 09 53 PM](https://github.com/user-attachments/assets/4f8d06d9-2160-45fc-a8b6-dd0d80a76dd9)

But on the domain invitations it shows this:
![Screenshot 2024-12-31 at 3 10 12 PM](https://github.com/user-attachments/assets/f7d79a5d-e43f-4786-a2b2-719c0115e5cb)


Mind just double checking this isn't from something you changed (or maybe a function that should be just reused in these two spots too). If this isn't from your code, and you want this to be a follow on ticket, I am good with that.  


I looked at your solution using concatenation, I would normally not be a fan of concatenating and then splitting due to it being slower to process. That said, most groups don't have much more than 10 domains total, so I doubt many users would see any slow down with that data size. I'm good with the solution for now as is, appreciate you fixing the bug.

rachidatecs commented:
> **Conditional approval:** merge after taking a quick look at the bug
> 
> **minor bug** I don't _think_ this is from your code, but there is a bug.
> 
> ...

Totally my code. Check the latest commit that fixes this. I also sneaked in an extra portfolio permission in fixtures (and improved one of the unit tests)

abroddrick commented on src/package-lock.json:
```diff
@@ -8888,13 +8845,15 @@
-      "resolved": "https://registry.npmjs.org/semver/-/semver-6.3.1.tgz",
-      "integrity": "sha512-BR7VvDCVHO+q2xBEWskxS6DJE1qRnb7DxzUrogb71CWoSficBxYsiAGd+Kl0mmq/MprG9yArRkyrQxTO6XjMzA==",
-      "dev": true,
+      "version": "7.6.3",
```
**non blocking question** Not sure if I noticed this change earlier, why was this updated? (just curious)

abroddrick commented:
Hate to say it, I'm still seeing this bug on the Member profile page
![image](https://github.com/user-attachments/assets/f51a8ecf-1fc0-4348-8c21-22e0cbf91535)
vs when I click manage on the domains I have 20 domains that I manage
![image](https://github.com/user-attachments/assets/2c4af64d-d27e-4a87-b6cf-80f88e288b7f)

abroddrick reviewed (approved):
Ignore my last comment, the bug was from bad data. Everything looks good now! Free to merge
